File: raytrace/src/lights.rs
use crate::tuple::Tuple;
use crate::color::Color;

#[derive(Debug)]
pub struct Light {
    pub position: Tuple,
    pub intensity: Color,
}

impl Light {
    pub fn equals(&self, l: Light) -> bool {
        self.position.equals( l.position ) &&
        self.intensity.equals( l.intensity )
    }
}

pub fn point_light(position: Tuple, intensity: Color) -> Light {
    Light { position: position, intensity: intensity }
}


#[cfg(test)]
mod tests {
    use crate::color::color;
    use crate::tuple::point;
    use crate::lights::point_light;

    #[test]
    fn point_light_has_position_and_intensity(){
        let position = point(0.0, 0.0, 0.0);
        let intensity = color(1.0, 1.0, 1.0);
        let light = point_light(position, intensity);
        assert!( light.position.equals( position ));
        assert!( light.intensity.equals( intensity ));
    }
}

File: python/lights.py
# to run tests: python -m unittest -v lights

import unittest
from tuples import point
from colors import color, WHITE

class Light:
    def __init__(self, position, intensity):
        self.position = position
        self.intensity = intensity

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return (self.position == other.position and
                    self.intensity == other.intensity)
        else:
            return False

def point_light(position, intensity):
    return Light(position, intensity)

class LightTestCase(unittest.TestCase):
    def test_a_point_light_has_a_position_and_intensity(self):
        position = point(0, 0, 0)
        intensity = WHITE
        
        light = point_light(position, intensity)

        self.assertEqual(light.position, point(0, 0, 0))
        self.assertEqual(light.intensity, WHITE)

# ---------------------------------------------------------------------------
if __name__ == '__main__':
    unittest.main()
